### ISSUE TITLE
Egress to port 443 denied due to CIDR check, but there is no CIDR rule in the policy

### ISSUE THREAD

**Issue by ashwinp** [cilium/cilium#3427]
### Title

Egress to port 443 denied due to CIDR, but there is no CIDR rule in the policy

### General Information

- Cilium version (run `cilium version`)
`v0.13.20`

- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
k8s

### How to reproduce the issue

Relevant rule:
```
- endpointSelector:
    matchLabels:
      "k8s:app": cloudwatch-exporter
      "k8s:io.kubernetes.pod.namespace": monitoring
  # Cloudwatch exporter talks to AWS API on port 443
  egress:
  - toPorts:
    - ports:
      - port: "443"
        protocol: TCP
```
From `cilium monitor` from the node running cloudwatch-exporter:
```
xx drop (Policy denied (CIDR)) flow 0x8acf7aee to endpoint 0, identity 29016->0: 172.203.118.32:42216 -> 54.240.254.162:443 tcp SYN
```

**Comment by tgraf** [comment]
Likely a dup of https://github.com/cilium/cilium/issues/3414

**Comment by ashwinp** [comment]
[lxcconf.txt](https://github.com/cilium/cilium/files/1864797/lxcconf.txt)
LXC config file (initial string has been snipped).

**Comment by joestringer** [comment]
I was suspecting it's due to these:

https://github.com/cilium/cilium/blob/82fbce193a79eea9c82e7a4f57de24375b5423e1/bpf/lib/maps.h#L239-L240

https://github.com/cilium/cilium/blob/82fbce193a79eea9c82e7a4f57de24375b5423e1/bpf/lib/maps.h#L253-L254

There's no CIDR policy defined in this case, but egress policy is enabled (for L4). However we're denying due to the CIDR. I suspect that userspace will not define `CIDR6_EGRESS_MAP` if there are no CIDR policies, and we'll hit this (and drop due to CIDR).

#3414, on the other hand, requires L7 to be involved.

**Comment by joestringer** [comment]
@tgraf we conversed about the following fix offline, but it some other issues will remain.

```
diff --git a/pkg/policy/api/egress.go b/pkg/policy/api/egress.go
index 21f8c83d131f..2771d6346176 100644
--- a/pkg/policy/api/egress.go
+++ b/pkg/policy/api/egress.go
@@ -113,3 +113,13 @@ type EgressRule struct {
 func (e *EgressRule) GetDestinationEndpointSelectors() EndpointSelectorSlice {
        return append(e.ToEndpoints, e.ToEntities.GetAsEndpointSelectors()...)
 }
+
+// AllL3Wildcarded returns true if the L3 selectors in the rule are all empty,
+// ie it applies to traffic from any L3 source.
+func (e *EgressRule) AllL3Wildcarded() bool {
+       return len(e.ToEndpoints) == 0 &&
+               len(e.ToCIDR) == 0 &&
+               len(e.ToCIDRSet) == 0 &&
+               len(e.ToEntities) == 0 &&
+               len(e.ToServices) == 0
+}
diff --git a/pkg/policy/rule.go b/pkg/policy/rule.go
index 4706cdf1e930..84df7385eba0 100644
--- a/pkg/policy/rule.go
+++ b/pkg/policy/rule.go
@@ -297,6 +297,10 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
                        }
                }
 
+               if egressRule.AllL3Wildcarded() {
+                       allCIDRs = append(allCIDRs, api.CIDRMatchAll...)
+               }
+
                if cnt := mergeCIDR(ctx, "Egress", allCIDRs, r.Labels, &result.Egress); cnt > 0 {
                        found += cnt
                }

```

In this case, it will address this issue but will not fix #3371, because the L4 check is still made first and the packet is dropped before CT, routing, etc in that case, so the CIDR allow doesn't get a chance to be checked.

I'm proposing instead a datapath approach to fix this issue as well as #3371 (See #3486).
